ros/src/computing/planning/common/lib/openplanner/op_planner/src/DecisionMaker.cpp

```diff
@@ -0,0 +1,503 @@
+		 m_TotalOriginalPath = globalPath;
+	 }
+ }
+
+ bool DecisionMaker::SelectSafeTrajectory()
+ {
+	 bool bNewTrajectory = false;
+	 PlannerHNS::PreCalculatedConditions *preCalcPrams = m_pCurrentBehaviorState->GetCalcParams();
+
+	 if(!preCalcPrams || m_RollOuts.size() == 0) return bNewTrajectory;
+
+	int currIndex = PlannerHNS::PlanningHelpers::GetClosestNextPointIndexFast(m_Path, state);
+	int index_limit = 0;
+	if(index_limit<=0)
+		index_limit =  m_Path.size()/2.0;
+	if(currIndex > index_limit
+			|| preCalcPrams->bRePlan
+			|| preCalcPrams->bNewGlobalPath)
+	{
+		//std::cout << "New Local Plan !! " << currIndex << ", "<< preCalcPrams->bRePlan << ", " << preCalcPrams->bNewGlobalPath  << std::endl;
+
+		preCalcPrams->bRePlan = false;
+		preCalcPrams->bNewGlobalPath = false;
+		bNewTrajectory = true;
+		m_Path = m_RollOuts.at(preCalcPrams->iCurrSafeTrajectory);
+	}
+
+	return bNewTrajectory;
+ }
+
+ PlannerHNS::BehaviorState DecisionMaker::GenerateBehaviorState(const PlannerHNS::VehicleState& vehicleState)
+ {
+	PlannerHNS::PreCalculatedConditions *preCalcPrams = m_pCurrentBehaviorState->GetCalcParams();
+
+	m_pCurrentBehaviorState = m_pCurrentBehaviorState->GetNextState();
+	if(m_pCurrentBehaviorState==0)
+		m_pCurrentBehaviorState = m_pInitState;
+
+	PlannerHNS::BehaviorState currentBehavior;
+
+	currentBehavior.state = m_pCurrentBehaviorState->m_Behavior;
+	currentBehavior.followDistance = preCalcPrams->distanceToNext;
+
+	if(preCalcPrams->bUpcomingRight)
+		currentBehavior.indicator = PlannerHNS::INDICATOR_RIGHT;
+	else if(preCalcPrams->bUpcomingLeft)
+		currentBehavior.indicator = PlannerHNS::INDICATOR_LEFT;
+	else
+		currentBehavior.indicator = PlannerHNS::INDICATOR_NONE;
+
+	currentBehavior.minVelocity		= 0;
+	currentBehavior.stopDistance 	= preCalcPrams->distanceToStop();
+	currentBehavior.followVelocity 	= preCalcPrams->velocityOfNext;
+	if(preCalcPrams->iPrevSafeTrajectory<0 || preCalcPrams->iPrevSafeTrajectory >= m_RollOuts.size())
+		currentBehavior.iTrajectory		= preCalcPrams->iCurrSafeTrajectory;
+	else
+		currentBehavior.iTrajectory		= preCalcPrams->iPrevSafeTrajectory;
+
+	return currentBehavior;
+ }
+
+ double DecisionMaker::UpdateVelocityDirectlyToTrajectory(const BehaviorState& beh, const VehicleState& CurrStatus, const double& dt)
+ {
+	if(m_TotalOriginalPath.size() ==0 ) return 0;
+
+	RelativeInfo info, total_info;
+	PlanningHelpers::GetRelativeInfo(m_TotalOriginalPath.at(m_iCurrentTotalPathId), state, total_info);
+	PlanningHelpers::GetRelativeInfo(m_Path, state, info);
+	double average_braking_distance = -pow(CurrStatus.speed, 2)/(m_CarInfo.max_deceleration) + m_params.additionalBrakingDistance;
+	double max_velocity	= PlannerHNS::PlanningHelpers::GetVelocityAhead(m_TotalOriginalPath.at(m_iCurrentTotalPathId), total_info, total_info.iBack, average_braking_distance);
+
+	unsigned int point_index = 0;
+	double critical_long_front_distance = m_CarInfo.length/2.0;
+
+
+	if(m_Path.size() <= 5)
```

Comment:
parametrize

Reply:
m_CarInfo.length is a parameter

Reply:
will think about better solution than this, added to the changing plan.